### PR TITLE
Unify and simplify interface of `Structure` and `StructureOperations`

### PR DESCRIPTION
**Pull request by t-reents**
As described in #68, I implemented a few changes to simplify the interface between `Structure` and `StructureOperations`.

The PR contains a Jupyter notebook showing that the desired interface actually works.

One thing to discuss: To keep the previous behavior, I had to introduce a different handling of single key/label indices. More precisely, if `append_to_coll == True` and you select a single key, a `StructureOperations` object is returned. However, if `append_to_coll == False`, the same indexing would simply return the single `Structure` object.

On the one hand, this might be considered as slightly inconsistent. On the other hand, it does actually make sense to only return the Structure if you don't want to add it to the initial `StructureCollection`. In case you want to add it, one can argue that the returned `StructureOperations` object is indeed the correct object to reflect it.

Comments are more than welcome.

### CODE REVIEW

**Comment by hdsassnick** [comment]
Sorry @t-reents, I am having a quite important deadline coming up on the 11th September. I will try to have a look as soon as possible but not sure if I can make it before that date...

File: aim2dat/strct/structure_operations.py
Comment (hdsassnick) [inline aim2dat/strct/structure_operations.py:1]: ``len(self)`` would need to be replaced with ``len(self.structures)``.

**Review by hdsassnick — CHANGES_REQUESTED** [review]
Hi @t-reents, just checked the new implementation and it looks good.

Two things I am not so sure about:

- Would it make sense to implement ``StructureOperations`` as a function? The problem right now is that it kind of is a wrapper for ``StructureOperations``. It is maybe not so straight-forward to do changes or get properties within ``StructureOperations`` on the ``StructureCollection`` object. I guess there are pros and cons for both ways, but maybe we can think about "class vs function" at this point before changing the interface.
- I think we should include this interface change in v0.2.0 (even if it means that this would postpone the new release) to be done with all the breaking changes.

**Comment by t-reents** [comment]
Hi @hdsassnick no worries regarding your deadline, please take your time. Thanks for giving it a look nonetheless!

How would the interface look like when using a function? How do you imagine to call a specific method on the underlying structures etc.?

Something like the following where the method is specified as the name (str)?
```
structure_operations_func(
    strct_coll,
    method_name,
    **dwarfs
)
```


I agree to think about those changes before doing a new release, to avoid breaking changes again and again.

**Comment by hdsassnick** [comment]
> Hi @hdsassnick no worries regarding your deadline, please take your time. Thanks for giving it a look nonetheless!
> 
> How would the interface look like when using a function? How do you imagine to call a specific method on the underlying structures etc.?
> 
> Something like the following where the method is specified as the name (str)?
> 
> ```
> structure_operations_func(
>     strct_coll,
>     method_name,
>     **dwarfs
> )
> ```
> 
> I agree to think about those changes before doing a new release, to avoid breaking changes again and again.

Yes, something like that with individual functions for manipulation, analysis, comparison and identification/removal of duplicate structures. The advantage would be that we are not "wrapping" ``StructureCollection``, so we do not have to care about exposing an interface to it. On the other hand, class properties are not available. Not sure if there are any other downsides to the function approach.

**Comment by t-reents** [comment]
Personally, I find the class approach more intuitive, in the sense that we provide an interface that allows to work with multiple structures in the same way as one would do with a single `Structure`.

One thing that would simplify the logic, at least in my opinion, is removing the logic of appending to the initial `StructureCollection`. In case of analysis methods, we are fine anyway, since they don't produce new structures. However, when manipulating structures, we would simply return a `StructureCollection` containing these results. The user would need to take care of how to combine it with existing data, which should be fine. 

Moreover, when working with a fixed set of structures, I think it feels natural to have this interface that allows to directly perform different operations on it. When using functions, you would always need to pass the same underlying dataset.
Finally, when removing the `append` logic, I think that we don't rely that much on `StructureCollection` which would make the logic quite simple. We only use it as the internal container that is iterated and provide access to the methods of the `Structure` objects.

**Comment by hdsassnick** [comment]
yes, sounds good. Then let us go for the class implementation without ``append`` functionality.

**Comment by hdsassnick** [comment]
Hi @t-reents, just wanted to quickly check on the status of the PR. Would be also happy to help with the implementation in case you are currently busy...

**Comment by t-reents** [comment]
Hi @hdsassnick, thanks for pinging me. Indeed, I was a bit busy with other things, but I picked up the work on this PR this morning. I basically finished the implementation (removing the append logic and small refactoring). I've also fixed the majority of tests. Only a few tests are failing due to my small refactoring. I'll try to fix them this evening.

In case I don't manage to fix those tests, I'll let you know.

Sorry for the delay!

**Comment by t-reents** [comment]
Fixed the tests. You can have another look now, @hdsassnick

**Comment by hdsassnick** [comment]
> Fixed the tests. You can have another look now, @hdsassnick

Hi @t-reents, I believe you forgot to push the commit?

**Comment by t-reents** [comment]
Oh wow 🤣 I also pushed it now

**Comment by hdsassnick** [comment]
Hi @t-reents ,

while trying to fix the ``ml`` sub-package, I encountered an issue:

The output of applying an operation on one structure divers to the application on several structures, as the function is called on the ``Structure`` object itself. I guess this is intended, but it makes the interface more cumbersome as the two cases need to be distinguished. I am wondering whether it would be more intuitive to always return a dictionary/pandas data frame regardless of the length of the key list?

The individual output can still be accessed in any case via: `strct_op.structures[key].function()`

What do you think?

**Comment by t-reents** [comment]
Totally agree. I was also thinking about it while working on this PR. For the historical context, the idea was that when you have a `StructureOperations` object but do the desired operation only on a single structure, the user could expect to get the same output as if it has been done on a "pure" `Structure` object. 

However, one can also argue that a different output from a single object, namely `StructureOperations`, might be even more confusing. Thinking about it now, I would also say that it is more important to have a uniform output for `StructureOperations`, as it simplifies the usage and also development of methods that rely on it. 

I'll fix it in this PR (if you agree?), so basically removing the following parts:

https://github.com/aim2dat/aim2dat/blob/d435beb7e38700c47c2c1d2f2889cd99cad0875e/aim2dat/strct/structure_operations.py#L957-L958

https://github.com/aim2dat/aim2dat/blob/d435beb7e38700c47c2c1d2f2889cd99cad0875e/aim2dat/strct/structure_operations.py#L963-L964

**Comment by hdsassnick** [comment]
That would be great, thank you.

I can then fix the remaining tests and I would also double-check and adapt the documentation.

**Comment by t-reents** [comment]
Fixed it.

Ah true, didn't think about the docs, and also forgot to check the other tests, I was only focussing on the `strct` module. Sorry! If you have the time to fix it, great, but otherwise, please let me know. Just to avoid that we are both working on the same.

**Review by hdsassnick — APPROVED** [review]
Hi @t-reents,

as promised, I fixed the tests (the Materials Project tests are probably failing due to the missing keys that are supplied via github secrets) and adjusted the documentation.

If you agree I would merge this PR and it should also close issue #70 and #68, right?

Closes #68

And it should also pave the path towards release v0.2.0. Please let me know if there are any other urgent changes that should be considered....

@Scemp, please let me know if there any objections from your side..

**Comment by hdsassnick** [comment]
Closes #68

**Comment by t-reents** [comment]
Thanks a lot, @hdsassnick!

#70 is not affected by this PR. I only tested the parallelization of the comparison methods in a branch of my fork. If you would like to include those changes in the new release as well, I could try to wrap it up today or tomorrow. In principle, I would only need to extend it to the other comparison methods. I only checked it for the f-fingerprints at the moment.

I'd suggest to do it in a separate PR and merge this one first.

**Pull request merged**